File: Java/Software Development Module - Java/Week13/TemperatureApp.java
/*
TemperatureApp.java
Billy Cussen
27/07/2020
*/

import javax.swing.JOptionPane;
import java.util.InputMismatchException;

public class TemperatureApp{

    public static void main(String [] args){

        //Variables
        String[] temperatureWeek = new String[3];
        String restart;
        String result;
        boolean inputCheckTemperature;
        
        //Object
        Temperature myTemperature;

        //Input
        do{
            restart = new String();

            for(int i = 0; i < temperatureWeek.length; i++){

                do{
                    inputCheckTemperature = true;
                    int commaCheck = 0;

                    try{
                        temperatureWeek[i] = JOptionPane.showInputDialog("Please enter Temperature for Week "+(i+1)+" in the following format:\n1,2,3,4,5,6,7");
                        temperatureWeek[i] = temperatureWeek[i].replaceAll("\\s+","");

                        for(int j = 0; j < temperatureWeek[i].length();j++){
                            if(temperatureWeek[i].charAt(j) == ','){
                                commaCheck++;
                            }
                        }

                        //https://stackoverflow.com/questions/37047386/check-if-string-contain-only-numbers-and-comma-in-java
                        if(!(temperatureWeek[i].matches("[0-9, /,]+")&&commaCheck == 6)){
                            throw new InputMismatchException();
                        }

                        //https://www.geeksforgeeks.org/split-string-java-examples/
                        String[] weekCheck = temperatureWeek[i].split(",");

                        for(String day : weekCheck){
                            if(day.equals("")){
                                throw new InputMismatchException();
                            }
                        }

                        //System.out.println("Comma Count, Week "+(i+1)+": "+commaCheck);
                        //System.out.println("Correct Format, Week "+(i+1)+": "+temperatureWeek[i].matches("[0-9, /,]+"));

                    }catch(InputMismatchException e){
                        JOptionPane.showMessageDialog(null,"Please enter this Weeks Temperatures in the following Format\n1,2,3,4,5,6,7");
                        inputCheckTemperature = false;
                    }

                }while(!inputCheckTemperature);

            }

            //Process
            myTemperature = new Temperature(temperatureWeek);
            result = myTemperature.getResult();
            myTemperature = null;
            
            //Output
            JOptionPane.showMessageDialog(null,result);

            //Check if the User wants to Retry
            do{

                try{
                    restart = JOptionPane.showInputDialog("Would you like to retry?\nY to Retry\nN to Exit");

                    if(!(restart.equalsIgnoreCase("Y")||restart.equalsIgnoreCase("N"))){
                        throw new InputMismatchException();
                    }

                } catch(InputMismatchException e){
                    JOptionPane.showMessageDialog(null,"You need to enter Y or N");
                }

            }while(!(restart.equalsIgnoreCase("Y")||restart.equalsIgnoreCase("N")));

        }while(restart.equalsIgnoreCase("Y"));

    }

}

File: Python/Software Development Module - Python/Week13/TemperatureApp.py
"""
TemperatureApp.py
Billy Cussen
19/10/2020
"""

#Objects
import math
import gc
import Temperature
from Temperature import myTemperature

#Variables
repeat = "y"

while repeat == "y":

    inputCheck = True

    while inputCheck:

        try:
            weeks = int(input("Please enter weeks to be checked (between 1 and 3): "))
            if weeks < 1 or weeks > 3:
                raise Exception
        except Exception as e:
            print("You need to enter a number between 1 and 3")
        else:
            if weeks >= 1 and weeks <= 3:
                inputCheck = False

    temperature = [[0 for j in range(7)] for i in range(weeks)] 

    for i in range(weeks):

        for j in range(7):

            while inputCheck == False:
                
                try:
                    temperature[i][j] = float(input("Please enter the Temperature for Week "+str(i+1)+", Day "+str(j+1)+": "))
                    if math.isnan(temperature[i][j]):
                        raise Exception
                except Exception as e:
                    print("You need to enter a valid number as the Temperature for Week "+str(i+1)+", day "+str(j+1)+"!")
                else:
                    inputCheck = True

            inputCheck = False

    #Process and Output
    myTemperatureApp = myTemperature(weeks,temperature)
    myTemperatureApp.getAverages()
    myTemperatureApp = None
    gc.collect()

    repeat = ""
    while repeat != "y" and repeat != "n":
        repeat = str(input("Please enter \"y\" to retry or \"n\" to exit: "))
        if repeat.lower() != "y" and repeat.lower() != "n":
            print("You need to enter \"y\" or \"n\"")

print("Thanks for using my Temperature App!")